packages/tailwind-typography/src/index.ts

```diff
@@ -0,0 +1,205 @@
+}
+
+export type TypographyOptionsVariants<TKeys extends readonly [string, ...string[]]> = Record<
+  string,
+  TypographyOptionsVariant | TypographyOptionsVariantMap<TKeys>
+>
+
+export type TypographyReturnVariant = Record<string, string | number>
+export type TypographyReturnVariantMap = Record<string, TypographyReturnVariant>
+export type TypographyReturnVariants = Record<
+  string,
+  TypographyReturnVariant | TypographyReturnVariantMap
+>
+
+export interface TypographyOptions<TKeys extends readonly [string, ...string[]]> {
+  breakpointKeys: TKeys
+  clamp?: boolean
+  clampMax?: boolean
+  clampMin?: boolean
+  fluid?: boolean
+  prefix?: string
+  unit?: string
+  variants: TypographyOptionsVariants<TKeys>
+}
+
+export function typography<TKeys extends readonly [string, ...string[]]>(
+  options: TypographyOptions<TKeys>,
+) {
+  const {
+    breakpointKeys,
+    clamp: globalClamp = false,
+    clampMax: globalClampMax = globalClamp,
+    clampMin: globalClampMin = globalClamp,
+    fluid: globalFluid = false,
+    prefix = '',
+    unit = 'px',
+    variants: variantsProp,
+  } = options
+
+  const isPlainVariant = (x: unknown): x is TypographyOptionsVariant =>
+    isPlainObject(x) && !(breakpointKeys[0] in x)
+
+  return plugin(({ addComponents, theme }) => {
+    const variants = Object.entries(variantsProp).reduce<TypographyReturnVariants>(
+      (acc, [variantNameProp, variantOrMap]) => {
+        // Compose the combined variant name.
+        const variantName = `.${prefix}${variantNameProp}`
+
+        // Compose the non-responsive variant.
+        if (isPlainVariant(variantOrMap)) {
+          const fontSize = `${variantOrMap.fontSize}${unit}`
+          acc[variantName] = { ...variantOrMap, fontSize }
+
+          return acc
+        }
+
+        const filteredBreakpointKeys = breakpointKeys.filter((key: TKeys[number]) =>
+          Boolean(variantOrMap[key]),
+        )
+
+        const getNextBreakpointKey = (key: TKeys[number]) => {
+          const index = filteredBreakpointKeys.indexOf(key)
+          return index !== -1 ? filteredBreakpointKeys[index + 1] : undefined
+        }
+
+        const {
+          clamp = globalClamp,
+          clampMax = clamp ?? globalClampMax,
+          clampMin = clamp ?? globalClampMin,
+          fluid = globalFluid,
+        } = variantOrMap
+        const firstBreakpointKey: TKeys[0] = breakpointKeys[0]
+        const firstVariant = variantOrMap[firstBreakpointKey]
+
+        // Compose the responsive variant map.
+        filteredBreakpointKeys.forEach((breakpoint: TKeys[number], idx) => {
+          const variant = variantOrMap[breakpoint]
+          const width: string = theme(`screens.${breakpoint}`)
+          const media = `@media (min-width: ${width})`
+
+          if (!fluid) {
+            const fontSize = `${variant.fontSize}${unit}`
+            let variantStyles = acc[variantName]
+
+            if (idx === 0) {
+              variantStyles = { ...variant, fontSize }
```

Comment:
**thought:** does this work?

I imagine that this will result in the following:

```
> const acc = {}
> let variantName = 'xs'
> let variantStyles = acc[variantName] // `undefined`
> variantStyles = {bar: true}
> acc[variantName] === variantStyles // `false`
```

Since `undefined` doesn't result in a reference to `acc`.

Reply:
Nice catch! Now fixed :)